### PR TITLE
[flow] add flow-typed libs + move to static immutable checks

### PR DESCRIPTION
**Pull request by jasonLaster**
Fixes Issue: #5122 

### Summary of Changes

* add flow-typed libs 
* removed immutable from immutable 
* I stubbed out `@babel/types` because it was causing new errors, guess we need to type this thing
* added `getSourceFromId` for cases where you know there is a source

### CODE REVIEW

**Comment by jasonLaster** [comment]
We now have a much higher coverage score

File: flow-typed/npm-custom/devtools-contextmenu_vx.x.x.js [hunk 1]
```diff
@@ -0,0 +1,7 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at <http://mozilla.org/MPL/2.0/>. */
+
+declare module "devtools-contextmenu" {
```
Comment (wldcordeiro) [inline flow-typed/npm-custom/devtools-contextmenu_vx.x.x.js]: An idea would be following this up with publishing these to `flow-typed`, but let's add them here first and flesh them out.

File: flow-typed/npm/babel-preset-env_vx.x.x.js [hunk 2]
```diff
@@ -0,0 +1,88 @@
+// flow-typed signature: 84e004258eec612c8ccb08e44dce7287
```
Comment (wldcordeiro) [inline flow-typed/npm/babel-preset-env_vx.x.x.js]: I don't really know if we need typedefs for most of the devDependencies, especially generated `any` stubs.

File: src/actions/ast.js [hunk 3]
```diff
@@ -51,13 +52,8 @@ export function setSourceMetaData(sourceId: SourceId) {
...
 export function setSymbols(sourceId: SourceId) {
   return async ({ dispatch, getState }: ThunkArgs) => {
-    const source = getSource(getState(), sourceId);
-    if (
-      !source ||
-      !source.text ||
-      source.isWasm ||
-      hasSymbols(getState(), source)
-    ) {
+    const source = getSourceFromId(getState(), sourceId);
```
Comment (wldcordeiro) [inline src/actions/ast.js]: Minor thing, storing the state as a variable saves a little from having to get it twice here.

File: src/actions/project-text-search.js [hunk 4]
```diff
@@ -58,12 +59,10 @@ export function searchSources(query: string) {
     await dispatch(addSearchQuery(query));
     dispatch(updateSearchStatus(statusType.fetching));
     const sources = getSources(getState());
-    const validSources = sources
-      .valueSeq()
-      .filter(
-        source =>
-          !hasPrettySource(getState(), source.id) && !isThirdParty(source)
-      );
+    const validSources = filter(
```
Comment (wldcordeiro) [inline src/actions/project-text-search.js]: If this is an array couldn't we do this and avoid lodash's `filter()`?
```js
sources.filter(source => !hasPrettySource(getState(), source.id) && !isThirdParty(source));
```

File: src/actions/sources/select.js [hunk 5]
```diff
@@ -104,8 +104,8 @@ export function selectLocation(location: Location) {
       return;
     }
 
-    const sourceRecord = getSource(getState(), location.sourceId);
-    if (!sourceRecord) {
+    const source = getSource(getState(), location.sourceId);
```
Comment (wldcordeiro) [inline src/actions/sources/select.js]: You had used the `getSourceById` in most of the code you changed but not here?

File: src/actions/tests/ui.spec.js [hunk 6]
```diff
@@ -15,7 +15,8 @@ const {
   getPaneCollapse,
   getHighlightedLineRange,
   getProjectDirectoryRoot,
-  getRelativeSources
+  getRelativeSources,
+  getSources
```
Comment (wldcordeiro) [inline src/actions/tests/ui.spec.js]: Unused?

File: src/reducers/sources.js [hunk 7]
```diff
@@ -239,19 +236,19 @@ function restoreTabs() {
...
  */
-function updateTabList(tabs: TabList, url: ?string, tabIndex?: number) {
+function updateTabList(tabs: TabList, url: ?string, tabIndex: ?number) {
   const urlIndex = tabs.indexOf(url);
   const includesUrl = !!tabs.find(tab => tab == url);
 
   if (includesUrl) {
     if (tabIndex != undefined) {
-      tabs = tabs.delete(urlIndex).insert(tabIndex, url);
+      // tabs = tabs.delete(urlIndex).insert(tabIndex, url);
```
Comment (wldcordeiro) [inline src/reducers/sources.js]: Commented code?

File: src/reducers/sources.js [hunk 8]
```diff
@@ -300,7 +297,10 @@ export function getNewSelectedSourceId(
       return "";
     }
 
-    const selectedSource = sources.find(source => source.url == selectedTabUrl);
+    const selectedSource = find(
```
[the remark this answers is not in the document]
Comment (jasonLaster) [reply src/reducers/sources.js]: in theory yes

File: src/reducers/sources.js (hunk 7 above)
Comment (jasonLaster) [reply src/reducers/sources.js]: yep, we should fix this :)

**Comment by darkwing** [comment]
How can we test this?  There's a ton going on :D

**Comment by jasonLaster** [comment]
haha - carefully :) it is not small

**Comment by jasonLaster** [comment]
selecting tabs is a bit funky

**Comment by darkwing** [comment]
All my testing went over great!  The only issue I spotted was that SourceTree icons aren't showing up; no "index" or "js" etc.

**Pull request merged**